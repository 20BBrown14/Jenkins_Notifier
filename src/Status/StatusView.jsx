import React from 'react';
import PropTypes from 'prop-types';
import Button from 'terra-button';
import Arrange from 'terra-arrange/lib/Arrange';
import './StatusView.scss';
import Text from 'terra-text';

const propTypes = {
  /** Repo information stored in app */
  repos: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  /** Button click handler for add new repo button */
  addRepoClickHandler: PropTypes.func.isRequired,
  /** Function to remove repo from list */
  removeRepo: PropTypes.func.isRequired,
};

const StatusView = (props) => {
  const {
    repos,
    addRepoClickHandler,
    removeRepo,
  } = props;

  const keys = [];
  if (repos) {
    Object.keys(repos).forEach((key) => {
      keys.push(key);
    });
  }

  /*
  for (let key in repos) {
    if (!repos.hasOwnProperty(key)) continue;
    keys.push(key);
  }
  */

  const content = [];
  for (let i = 0; i < keys.length; i += 1) {
    const arrangeFill = (
      <Text
        fontSize={16}
      >
        {keys[i]}
      </Text>);

    const arrangeStart = (
      <div>
        <Button
          key={`${keys[i]} start Jenkins`}
          className="arrangeButton"
          text="View on Jenkins"
          onClick={() => { window.open(repos[keys[i]].URL); }}
        />
        <Button
          key={`${keys[i]} start jobs`}
          className="arrangeButton"
          text="View Jobs"
          onClick={() => {}}
        />
      </div>
    );

    const arrangeEnd = (
      <Button
        key={`${keys[i]} end`}
        className="arrangeButton"
        text="Remove Repo"
        onClick={() => { removeRepo(`${keys[i]}`); }}
      />
    );
    content.push(
      <Arrange
        key={`${keys[i]} fill`}
        className="statusArrange"
        fitStart={arrangeStart}
        alignFitStart="center"
        fill={arrangeFill}
        alignFill="center"
        fitEnd={arrangeEnd}
        alignFitEnd="center"
      />,
    );
  }
  // content.push(<Button style={{ margin: '0px 5px 0px 5px' }} text="Confirm" />);

  return (
    <div>
      <div className="StatusView-Window">
        {content}
      </div>
      <Button
        style={{ margin: '12.5px 5px 0px 5px' }}
        text="Add New Repo"
        onClick={addRepoClickHandler}
      />
    </div>
  );
};

StatusView.propTypes = propTypes;
export default StatusView;
