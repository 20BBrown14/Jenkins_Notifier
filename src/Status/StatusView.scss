.StatusView-Window {
  width: 650px;
  height: 370px;
  overflow-y: auto;
  border-bottom: inset;
  border-bottom-color: lightgray;
  border-width: 2px;
}

.statusArrange {
  border-bottom: solid;
  border-width: 2px;
  border-color: lightgray;
}

.arrangeButton {
  margin: 0px;
}

.center {
  margin: auto;
  width: 100%;
  padding: 5px;
}

.titleText {
  margin: auto;
  width: 100%;
  padding: 5px;
  border-bottom: solid;
  border-width: 2px;
  border-color: lightgray;
  text-align: center;
}

.textCenter {
  text-align: center;
}

.tooltip {
  position: relative;
  display: inline-block;
}

.tooltip .tooltiptext {
  visibility: hidden;
  width: 120px;
  background-color: #808080;
  text-align: center;
  border-radius: 6px;
  padding: 5px 0;
  position: absolute;
  z-index: 1;
  top: 85%;
  opacity: 0;
  transition: opacity 0.5s;
  transition-delay: 0.75s;
}

.tooltip:hover .tooltiptext{
  visibility: visible;
  opacity: 1;
}

.refreshButton {
  top: -22px;
  right: 5px;
  position: absolute;
}

.addRepoButton {
  position: relative;
  float: right;
  top: 5px;
}