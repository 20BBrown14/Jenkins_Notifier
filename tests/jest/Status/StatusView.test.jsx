import React from 'react';
import StatusView from '../../../src/Status/StatusView';
import determineImage from '../../../src/modules/jobStatus';

jest.mock('../../../src/modules/jobStatus');

describe('StatusView', () => {
  it('should render a default view', () => {
    const repoInformation = {
      someKey: 'someValue',
      anotherKey: 'anotherValue',
    };
    const testView = (shallow(
      <StatusView
        repos={repoInformation}
        addRepoClickHandler={() => {}}
        removeRepo={() => {}}
        viewJobs={() => {}}
        removeJob={() => {}}
      />,
    ));
    expect(testView).toMatchSnapshot();
  });
  describe('Repos', () => {
    it('should render a view', () => {
      const repoInformation = {
        someRepo: {
          url: 'someUrl',
          jobs: {
            firstKeya: {
              name: 'firstKeya',
            },
            secondKeya: {
              name: 'secondKeya',
            },
          },
        },
        anotherRepo: {
          url: 'anotherUrl',
          jobs: {
            firstKeyb: {
              name: 'firstKeyb',
            },
            secondKeyb: {
              name: 'secondKeyb',
            },
          },
        },
        andAnotherREpo: {
          url: 'andAnotherUrl',
          jobs: {
            firstKeyc: {
              name: 'firstKeyc',
            },
            secondKeyc: {
              name: 'secondKeyc',
            },
          },
        },
      };
      const testView = (mount(
        <StatusView
          repos={repoInformation}
          addRepoClickHandler={() => {}}
          noRepos={() => {}}
          removeRepo={() => {}}
          viewJobs={() => {}}
          removeJob={() => {}}
        />,
      ));
      expect(testView).toMatchSnapshot();
      expect(testView.find('Button').length).toEqual(10);
      expect(testView.find('Arrange').length).toEqual(3);
    });
  });

  describe('Jobs', () => {
    it('should render a view', () => {
      determineImage.mockImplementation(color => (color === 'blue' ? 'successful' : 'unknown_status'));
      const repoInformation = {
        someRepo: {
          url: 'someUrl',
          jobs: {
            firstKeya: {
              name: 'firstKeya',
              color: 'blue',
            },
            secondKeya: {
              name: 'secondKeya',
            },
          },
        },
        anotherRepo: {
          url: 'anotherUrl',
          jobs: {
            firstKeyb: {
              name: 'firstKeyb',
            },
            secondKeyb: {
              name: 'secondKeyb',
            },
          },
        },
      };
      const repoToView = 'someRepo';
      const testView = (mount(
        <StatusView
          repos={repoInformation}
          addRepoClickHandler={() => {}}
          noRepos={() => {}}
          removeRepo={() => {}}
          viewJobs={() => {}}
          repoToView={repoToView}
          removeJob={() => {}}
        />,
      ));
      expect(testView).toMatchSnapshot();
      expect(testView.find('Button').length).toEqual(7);
      expect(testView.find('Arrange').length).toEqual(2);
    });
  });
  describe('Error', () => {
    it('should render an error view', () => {
      const testView = (mount(
        <StatusView
          repos={undefined}
          addRepoClickHandler={() => {}}
          noRepos={() => {}}
          removeRepo={() => {}}
          viewJobs={() => {}}
          repoToView={undefined}
          removeJob={() => {}}
        />,
      ));
      expect(testView).toMatchSnapshot();
    });
  });

  describe('Repo Click handlers', () => {
    let repoInformation;
    let testView;
    let viewJobs;
    let removeRepo;
    beforeEach(() => {
      repoInformation = {
        someRepo: {
          URL: 'someUrl',
          jobs: {
            firstKeya: {
              name: 'firstKeya',
            },
            secondKeya: {
              name: 'secondKeya',
            },
          },
        },
        anotherRepo: {
          URL: 'anotherUrl',
          jobs: {
            firstKeyb: {
              name: 'firstKeyb',
            },
            secondKeyb: {
              name: 'secondKeyb',
            },
          },
        },
        andAnotherRepo: {
          URL: 'andAnotherUrl',
          jobs: {
            firstKeyc: {
              name: 'firstKeyc',
            },
            secondKeyc: {
              name: 'secondKeyc',
            },
          },
        },
      };
      viewJobs = jest.fn();
      removeRepo = jest.fn();
      testView = (mount(
        <StatusView
          repos={repoInformation}
          addRepoClickHandler={() => {}}
          noRepos={() => {}}
          removeRepo={removeRepo}
          viewJobs={viewJobs}
          removeJob={() => {}}
        />,
      ));
    });
    describe('View Repo Button', () => {
      it('should open jenkins window', () => {
        global.open = jest.fn();
        expect(testView).toMatchSnapshot();
        const buttons = testView.find('Button');
        expect(buttons.length).toEqual(10);
        buttons.at(0).simulate('click');
        expect(global.open).toHaveBeenCalledTimes(1);
        expect(global.open.mock.calls[0][0]).toEqual('someUrl');
        buttons.at(3).simulate('click');
        expect(global.open).toHaveBeenCalledTimes(2);
        expect(global.open.mock.calls[1][0]).toEqual('anotherUrl');
        buttons.at(6).simulate('click');
        expect(global.open).toHaveBeenCalledTimes(3);
        expect(global.open.mock.calls[2][0]).toEqual('andAnotherUrl');
      });
    });
    describe('View Jobs Button', () => {
      it('should call view jobs prop function', () => {
        expect(testView).toMatchSnapshot();
        const buttons = testView.find('Button');
        expect(buttons.length).toEqual(10);
        buttons.at(1).simulate('click');
        expect(viewJobs).toHaveBeenCalledTimes(1);
        expect(viewJobs.mock.calls[0][0]).toEqual('someRepo');
        buttons.at(4).simulate('click');
        expect(viewJobs).toHaveBeenCalledTimes(2);
        expect(viewJobs.mock.calls[1][0]).toEqual('anotherRepo');
        buttons.at(7).simulate('click');
        expect(viewJobs).toHaveBeenCalledTimes(3);
        expect(viewJobs.mock.calls[2][0]).toEqual('andAnotherRepo');
      });
    });
    describe('Remove repo button', () => {
      it('should call remove repo prop function', () => {
        expect(testView).toMatchSnapshot();
        const buttons = testView.find('Button');
        expect(buttons.length).toEqual(10);
        buttons.at(2).simulate('click');
        expect(removeRepo).toHaveBeenCalledTimes(1);
        expect(removeRepo.mock.calls[0][0]).toEqual('someRepo');
        buttons.at(5).simulate('click');
        expect(removeRepo).toHaveBeenCalledTimes(2);
        expect(removeRepo.mock.calls[1][0]).toEqual('anotherRepo');
        buttons.at(8).simulate('click');
        expect(removeRepo).toHaveBeenCalledTimes(3);
        expect(removeRepo.mock.calls[2][0]).toEqual('andAnotherRepo');
      });
    });
  });

  describe('Job Click handlers', () => {
    let repoInformation;
    let repoToView;
    let testView;
    let removeJob;
    beforeEach(() => {
      repoInformation = {
        someRepo: {
          url: 'someUrl',
          jobs: {
            firstKeya: {
              name: 'firstKeya',
              color: 'blue',
              url: 'firstKeyUrla',
            },
            secondKeya: {
              name: 'secondKeya',
              url: 'secondKeyUrla',
            },
          },
        },
        anotherRepo: {
          url: 'anotherUrl',
          jobs: {
            firstKeyb: {
              name: 'firstKeyb',
              url: 'firstKeyUrlb',
            },
            secondKey: {
              name: 'secondKeyb',
              url: 'secondKeyUrlb',
            },
          },
        },
      };
      repoToView = 'someRepo';
      removeJob = jest.fn();
      testView = (mount(
        <StatusView
          repos={repoInformation}
          addRepoClickHandler={() => {}}
          noRepos={() => {}}
          removeRepo={() => {}}
          viewJobs={() => {}}
          repoToView={repoToView}
          removeJob={removeJob}
        />,
      ));
    });
    describe('View Job Button', () => {
      it('should call open window function', () => {
        global.open = jest.fn();
        expect(testView).toMatchSnapshot();
        const buttons = testView.find('Button');
        expect(buttons.length).toEqual(7);
        buttons.at(0).simulate('click');
        expect(global.open).toHaveBeenCalledTimes(1);
        expect(global.open.mock.calls[0][0]).toEqual('firstKeyUrla');
        buttons.at(2).simulate('click');
        expect(global.open).toHaveBeenCalledTimes(2);
        expect(global.open.mock.calls[1][0]).toEqual('secondKeyUrla');
      });
    });
    describe('Remove job button', () => {
      it('should call removeJob prop function', () => {
        expect(testView).toMatchSnapshot();
        const buttons = testView.find('Button');
        expect(buttons.length).toEqual(7);
        buttons.at(1).simulate('click');
        expect(removeJob).toHaveBeenCalled();
        expect(removeJob).toHaveBeenCalledTimes(1);
        expect(removeJob.mock.calls[0][0]).toEqual('firstKeya');
        expect(removeJob.mock.calls[0][1]).toEqual('someRepo');
        buttons.at(3).simulate('click');
        expect(removeJob).toHaveBeenCalledTimes(2);
        expect(removeJob.mock.calls[1][0]).toEqual('secondKeya');
        expect(removeJob.mock.calls[1][1]).toEqual('someRepo');
      });
    });
  });
});
