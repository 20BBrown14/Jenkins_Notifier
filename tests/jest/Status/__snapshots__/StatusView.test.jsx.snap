// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusView should render a default view 1`] = `
<div>
  <div
    className="StatusView-Window"
  >
    <Arrange
      alignFill="center"
      alignFitEnd="center"
      alignFitStart="center"
      className="statusArrange"
      fill={
        <Text
          color="inherit"
          fontSize={16}
          isItalic={false}
          isVisuallyHidden={false}
        >
          someKey
        </Text>
      }
      fitEnd={
        <Button
          className="arrangeButton"
          isBlock={false}
          isCompact={false}
          isDisabled={false}
          isIconOnly={false}
          isReversed={false}
          onClick={[Function]}
          text="Remove Repo"
          type="button"
          variant="neutral"
        />
      }
      fitStart={
        <div>
          <Button
            className="arrangeButton"
            isBlock={false}
            isCompact={false}
            isDisabled={false}
            isIconOnly={false}
            isReversed={false}
            onClick={[Function]}
            text="View on Jenkins"
            type="button"
            variant="neutral"
          />
          <Button
            className="arrangeButton"
            isBlock={false}
            isCompact={false}
            isDisabled={false}
            isIconOnly={false}
            isReversed={false}
            onClick={[Function]}
            text="View Jobs"
            type="button"
            variant="neutral"
          />
        </div>
      }
      key="someKey fill"
    />
    <Arrange
      alignFill="center"
      alignFitEnd="center"
      alignFitStart="center"
      className="statusArrange"
      fill={
        <Text
          color="inherit"
          fontSize={16}
          isItalic={false}
          isVisuallyHidden={false}
        >
          anotherKey
        </Text>
      }
      fitEnd={
        <Button
          className="arrangeButton"
          isBlock={false}
          isCompact={false}
          isDisabled={false}
          isIconOnly={false}
          isReversed={false}
          onClick={[Function]}
          text="Remove Repo"
          type="button"
          variant="neutral"
        />
      }
      fitStart={
        <div>
          <Button
            className="arrangeButton"
            isBlock={false}
            isCompact={false}
            isDisabled={false}
            isIconOnly={false}
            isReversed={false}
            onClick={[Function]}
            text="View on Jenkins"
            type="button"
            variant="neutral"
          />
          <Button
            className="arrangeButton"
            isBlock={false}
            isCompact={false}
            isDisabled={false}
            isIconOnly={false}
            isReversed={false}
            onClick={[Function]}
            text="View Jobs"
            type="button"
            variant="neutral"
          />
        </div>
      }
      key="anotherKey fill"
    />
  </div>
  <Button
    isBlock={false}
    isCompact={false}
    isDisabled={false}
    isIconOnly={false}
    isReversed={false}
    onClick={[Function]}
    style={
      Object {
        "margin": "12.5px 5px 0px 5px",
      }
    }
    text="Add New Repo"
    type="button"
    variant="neutral"
  />
</div>
`;
